Support workload log collection

inf17101 opened the pull request:
Issues: #9 

<!--  Description of the change in case no issue is mentioned -->

# Definition of Done

The PR shall be merged only if all items mentioned in [CONTRIBUTING.md](../CONTRIBUTING.md#how-to-contribute) have been followed. In case an item is not applicable as described, please provide a short explanation in the description.

- [ ] All steps in [CONTRIBUTING.md](../CONTRIBUTING.md#how-to-contribute) have been handled

inf17101 commented:
> I think I must add more tests to the coverage. I will check again.

Now, all tests are added. Review can be started.

krucod3 commented on src/components/control_interface.rs:
```diff
@@ -467,15 +656,18 @@ mod tests {
+        ankaios::CHANNEL_SIZE, components::workload_state_mod::WorkloadInstanceName, LogEntry,
+        LogResponse,
+    };
+    use crate::{ankaios_api, components::response::generate_test_proto_log_entries_response};
```
Strange that the use statements are spread out, this way...
There is already an entry for the `crate::components::response` on line 662-663

inf17101 replied:
Fixed in [0a34672](https://github.com/eclipse-ankaios/ank-sdk-rust/pull/22/commits/0a34672464cce597a38ef235728009b97d02525a)

krucod3 commented on src/components/response.rs:
```diff
@@ -285,6 +328,50 @@ pub fn generate_test_response_update_state_success(req_id: String) -> Response {
 }
 
+#[cfg(test)]
+pub fn generate_test_proto_log_entries_response(request_id: String) -> FromAnkaios {
```
Well that is actually a `FromAnkaios` message that is generated and not a response.

krucod3 replied:
Fixed.

krucod3 commented on src/components/response.rs:
```diff
@@ -157,6 +171,35 @@ impl From<FromAnkaios> for Response {
                                 UpdateStateSuccess::new_from_proto(update_state_success),
                             ))
                         }
+                        AnkaiosResponseContent::LogsRequestAccepted(logs_request_accepted) => {
+                            ResponseType::LogsRequestAccepted(
+                                logs_request_accepted
+                                    .workload_names
+                                    .into_iter()
+                                    .map(WorkloadInstanceName::from)
+                                    .collect(),
+                            )
+                        }
+                        AnkaiosResponseContent::LogsCancelAccepted(_) => {
+                            ResponseType::LogsCancelAccepted
+                        }
+                        AnkaiosResponseContent::LogEntriesResponse(log_entries_response) => {
+                            let log_entries = log_entries_response
+                                .log_entries
+                                .into_iter()
+                                .map(LogEntry::from)
+                                .collect();
+
+                            ResponseType::LogEntriesResponse(log_entries)
+                        }
+                        AnkaiosResponseContent::LogsStopResponse(logs_stop_response) => {
+                            let instance_name = logs_stop_response
+                                .workload_name
+                                .map(WorkloadInstanceName::from)
+                                .unwrap_or_unreachable();
+
+                            ResponseType::LogsStopResponse(instance_name)
+                        }
```
utest are missing for this.

inf17101 replied:
Fixed in [e9e38ea](https://github.com/eclipse-ankaios/ank-sdk-rust/pull/22/commits/e9e38eaae36aac71cd7f598420f0f56838afa349)

krucod3 reviewed (commented):
A couple of finding. Review not completed yet.

inf17101 commented on src/components/control_interface.rs:
```diff
@@ -65,6 +67,62 @@ pub enum ControlInterfaceState {
+    /// * `request_id` - A [String] representing the request ID of the initial logs request of the log campaign;
+    /// * `logs_sender` - A [`mpsc::Sender<LogResponse>`] to forward log responses for the log campaign.
+    ///
+    pub fn insert(&mut self, request_id: String, logs_sender: mpsc::Sender<LogResponse>) {
```
Fixed in [985726a](https://github.com/eclipse-ankaios/ank-sdk-rust/pull/22/commits/985726a62a540009b38ebc70d3d0e685d843ce31)

inf17101 commented on src/components/control_interface.rs:
```diff
@@ -1010,24 +1017,13 @@ mod tests {
+            log_entries_response
+                .log_entries
+                .into_iter()
+                .map(|entry| entry.into())
```
I have avoided to use the conversion here, since the test tests something different. Now we have the `from` coupled into the test. Are you sure we should keep that?
Before a clear list of expected entries were defined.

krucod3 replied:
The from is tested somewhere else explicitly. Theoretically we have coupling this way, on the other hand, we are generating something in a function and then defining the expectations locally. When a change is needed in the object and/or in the generate function, we would have more maintenance effort as all locally defined objects must be adapted. 

I think in this particular case the coupling is the better choice.

krucod3 commented on src/components/control_interface.rs:
```diff
@@ -772,4 +967,325 @@ mod tests {
+                ),
+                message: "log message 2".to_owned(),
+            },
+        ]);
```
Fixed.

inf17101 commented:
I have added the `LogsCancelAccepted` response handling and currently testing it in the test workload. Commit: [255d0c6](https://github.com/eclipse-ankaios/ank-sdk-rust/pull/22/commits/255d0c6a282b59f2ecaa843cb4fe2ae163a4887c)

inf17101 commented:
> I have added the `LogsCancelAccepted` response handling and currently testing it in the test workload. Commit: [255d0c6](https://github.com/eclipse-ankaios/ank-sdk-rust/pull/22/commits/255d0c6a282b59f2ecaa843cb4fe2ae163a4887c)

I reverted the commit with the `LogsCancelAccepted` message, because the response is only received, when the client code of the SDK does actively cancel the log campaign.

inf17101 commented on src/components/response.rs:
```diff
@@ -308,94 +441,280 @@ mod tests {
+            ))),
+        });
+
+        assert_eq!(response.get_request_id(), "123".to_owned());
+        assert_eq!(
+            format!("{}", response.get_content()),
+            format!("{}", ResponseType::LogsRequestAccepted(Vec::default()))
+        );
```
I will follow like discussed, I change the code to the better version and the existing utests in the same style we adapt within an extra issue.

inf17101 commented on src/components/response.rs:
```diff
@@ -308,94 +441,280 @@ mod tests {
+                "workload_B".to_owned(),
+                "id_b".to_owned()
+            )
+        );
```
Fixed in [2a6c60b](https://github.com/eclipse-ankaios/ank-sdk-rust/pull/22/commits/2a6c60b35b2e59950f6cf4b906aebb2934e1cc78)

krucod3 reviewed (approved):
LGTM 👍
